core-plugins/src/main/java/io/cdap/plugin/batch/joiner/Joiner.java

```diff
@@ -95,11 +95,15 @@ public void prepareRun(BatchJoinerContext context) {
       context.setNumPartitions(conf.getNumPartitions());
     }
     FailureCollector collector = context.getFailureCollector();
-    init(context.getInputSchemas(), collector);
-    collector.getOrThrowException();
-    Collection<OutputFieldInfo> outputFieldInfos = createOutputFieldInfos(context.getInputSchemas(), collector);
-    collector.getOrThrowException();
-    context.record(createFieldOperations(outputFieldInfos, perStageJoinKeys));
+    Map<String, Schema> inputSchemas = context.getInputSchemas();
+    // inputSchemas will be empty if the output schema of the previous node is a macro
+    if (!inputSchemas.isEmpty()) {
```

Comment:
nit: it's more readable in these situations to do:

```
if (inputSchemas.isEmpty()) {
  return;
}

// rest of logic
```

doesn't make much difference here, but is more true if the logic within the block is longer.

Reply:
the correct behavior for the app would not be to give an empty map, but to have a null value as the schema. Let's check for that scenario as well (there is at least one null in the map values), so that this doesn't need to be revisited once the getInputSchemas() method is fixed.

Reply:
fixed.